### ISSUE TITLE
Implementing multi-problem from scratch in t2t

### ISSUE THREAD

**Issue by stefan-falk** [tensorflow/tensor2tensor#1687]
I was hoping to get some help on a custom multi-problem implementation.

Basically I want to implement `multi_problem.MultiProblem` from scratch. Why? I want to remove the mandatory language-model problem as a first task and I also want to make it compatible with my own data-pipeline. We have a separate pipeline which generates the shards and our custom t2t-problems only load those shards (for single-problem training).

Now, since I was not able to modify `MultiProblem` s.t. it actually worked during training (the problem was that the loss was constantly 0), I decided to implement my own version. 

However, now that I got the first training running, I have the same problem: The loss is constantly 0. 

#### What am I doing?

Actually I have wrapped my _working_ problem implementation and all I do is I return the `Dataset` of the first task to tensor2tensor. 

#### How are you doing it?

Basically I have problems `a`, `b`, .. 
For each problem I create shards in separate directories:

```
a/shards/001-of-100.train.shard
         002-of-100.train.shard
         001-of-001.test.shard
b/shards/001-of-100.train.shard
         002-of-100.train.shard
         001-of-001.test.shard
...
```

Behind all this, there is a _configuration_ file which allows me to configure certain things. Those shards are then simply loaded using t2t - basically by overriding `Problem.filepattern`. That's (basically) it. No magic involved.

#### How does the multi-problem look like?

Essentially I am just wrapping `MyTranslation` problem into `MyTranslationTask`. It really is just a facade. `self._dataset` is this config-based thing which creates the shards. All this is not really important but nevertheless:

```python
class MyTranslationTask(MyTranslation):

    def __init__(
        self,
        was_reversed=False, was_copy=False,
        parent_id=None,
        config_fp=None, archive_root=None, datasets_root=None,
        shared_vocab_fp: str = None, source_lang=None, target_lang=None
    ):
        super().__init__(was_reversed=was_reversed, was_copy=was_copy)

        self._source_lang = source_lang
        self._target_lang = target_lang

        self._encoders = {
            'inputs': SubwordEncoder(shared_vocab_fp),
            'targets': SubwordEncoder(shared_vocab_fp)
        }

        self._dataset = Dataset(
            datasets_root,
            archive_root,
            config_fp=config_fp,
            encoders=self._encoders,
            parent_id=parent_id
        )

    def feature_encoders(self, data_dir):
        return self._encoders

    def generate_data(self, data_dir, tmp_dir, task_id=-1):
        self._dataset.create_shards()

    def filepattern(self, data_dir, split: str, shard=None):
        return super().filepattern(self._dataset.shards_dir, split, shard)

    @property
    def name(self):
        return '%s2%s_%s' % (self.source_lang, self.target_lang, super().name)

    @property
    def shards_dir(self):
        return self._dataset.shards_dir

    @property
    def source_lang(self):
        return self._source_lang

    @property
    def target_lang(self):
        return self._target_lang
```

This is my attempt for the custom `MultiProblem` aka `MyMultiProblem`. All it really does is load the list of tasks and return a dataset. 

To keep it simple, I only return the dataset of the first task. See `def dataset()` below.

```python
@registry.register_problem('my_multi_problem')
class MyMultiProblem(problem.Problem):

    def __init__(self, was_reversed=False, was_copy=False):
        super(MyMultiProblem, self).__init__(was_reversed, was_copy)
        # Mandatory field 'task_list'. Must contain a list of t2t problems
        self.task_list = []

    def generate_data(self, data_dir, tmp_dir, task_id=-1):
        self._initialize(data_dir)
        for task in self.task_list:
            task.generate_data(data_dir, tmp_dir, task_id)

    def get_hparams(self, model_hparams=None):
        logger = tf.logging
        if model_hparams:
            logger.info('[%s] - Setting hparams' % self.name)
            self._initialize(model_hparams.data_dir)
            thresholds = ','.join([str(task_cfg['threshold']) for task_cfg in self.config.tasks])
            model_hparams.multiproblem_per_task_threshold = '%s' % thresholds
            model_hparams.multiproblem_mixing_schedule = 'constant'

            self._hparams = self.task_list[0].get_hparams(model_hparams)
            # Increase the vocab size to account for task ids and modify the modality.
            vocab_size_inc = len(self.task_list)
            vocab_size_inc += self.get_max_num_classes()
            vocab_size = self._hparams.vocabulary['inputs'].vocab_size
            new_vocab_size = vocab_size + vocab_size_inc

            self.update_task_ids(vocab_size)
            self._hparams.vocab_size['inputs'] = new_vocab_size
            self._hparams.modality['inputs'] = modalities.SymbolModality(
                model_hparams, self._hparams.vocab_size['inputs']
            )

            self._hparams.vocab_size['targets'] = self._hparams.vocabulary['targets'].vocab_size
            self._hparams.modality['targets'] = modalities.SymbolModality(
                model_hparams, self._hparams.vocab_size['targets']
            )

            logger.info('[%s] - Old vocabulary size: %d' % (self.name, vocab_size))
            logger.info('[%s] - New vocabulary size: %d' % (self.name, new_vocab_size))
            return self._hparams

        if self._hparams:
            return self._hparams

        assert len(self.task_list) > 0, 'Cannot initialize problem hparams without data_dir'

    def eval_hooks(self, features, logits, hparams):
        return [MultiProblemEvalHook(features, logits, hparams)]

    def eval_metrics(self):
        return [
            metrics.Metrics.ACC,
            metrics.Metrics.NEG_LOG_PERPLEXITY,
            metrics.Metrics.APPROX_BLEU
        ]

    def filepattern(self, data_dir, mode, shard=None):
        tf.logging.info("Generating multi problem filepattern")
        return [task.filepattern(None, mode, shard) for task in self.task_list]

    def update_task_ids(self, encoder_vocab_size):
        logger = tf.logging
        for idx, task in enumerate(self.task_list):
            task.set_task_id(idx + encoder_vocab_size)
            logger.info('[%s] - Task %d (%s) has id %d.' % (self.name, idx, task.name, task.task_id))

    def get_max_num_classes(self):
        num = 0
        for task in self.task_list:
            if hasattr(task, "num_classes"):
                if num < task.num_classes:
                    num = task.num_classes
        return num

    def add_task_id(self, task, example, encoder, hparams, is_infer):
        concat_list = [[task.task_id], example['inputs']]
        example['inputs'] = tf.concat(concat_list, axis=0)
        example['task_id'] = tf.constant([task.task_id], dtype=tf.int64)
        return example

    def dataset(
        self,
        mode,
        data_dir=None,
        num_threads=None,
        output_buffer_size=None,
        shuffle_files=None,
        hparams=None,
        preprocess=True,
        dataset_split=None,
        shard=None,
        partition_id=0,
        num_partitions=1,
        shuffle_buffer_size=1024,
        max_records=-1
    ):

        is_training = mode == tf.estimator.ModeKeys.TRAIN
        is_infer = mode == tf.estimator.ModeKeys.PREDICT

        task = self.task_list[0]

        task_dataset = task.dataset(
            mode=mode,
            data_dir=self.task_list[0].shards_dir,
            num_threads=num_threads,
            output_buffer_size=output_buffer_size,
            shuffle_files=shuffle_files,
            hparams=hparams,
            preprocess=preprocess,
            dataset_split=dataset_split,
            shard=shard,
            partition_id=partition_id,
            num_partitions=num_partitions,
            shuffle_buffer_size=shuffle_buffer_size,
            max_records=max_records
        )

        task_dataset = task_dataset.map(lambda x: self.add_task_id(task, x, self.subword_encoder, hparams, is_infer))

        return task_dataset

    @property
    def name(self):
        return corpus_name(MyMultiProblem)

    @property
    def vocab_type(self):
        return VocabType.SUBWORD

    @property
    def num_generate_tasks(self):
        return 1

    def _initialize(self, data_dir):
        logger = tf.logging
        if self.task_list and len(self.task_list) > 0:
            return

        mp_config_fp = flags.FLAGS.multi_problem_config_fp

        with tf.gfile.Open(mp_config_fp) as f:
            self.config = Config(**yaml.full_load(f))

        # ...

        self.subword_encoder = self.get_subword_encoder()

        # Create task list

        self.task_list += multi_problem_util.create_tasks(
            config_fps=config_fps,
            allowed_pairs=language_pairs,
            datasets_root=datasets_root,
            shared_vocab_fp=self.subword_encoder.vocab_fp,
            task_id_offset=self.subword_encoder.vocab_size,
            parent_id=lm_problem.fingerprint
        )

        logger.info('Created task list:')
        for task in self.task_list:
            logger.info(' - %s' % task.name)
```

#### The problem

The problem is that the loss is still constantly 0 and I do not see what could be causing this. 

From the outside, everything is looking fine. I have two different vocabularies (as expected):

```
INFO:tensorflow:Transforming feature 'inputs' with symbol_modality_8772_1024.bottom
INFO:tensorflow:Transforming feature 'targets' with symbol_modality_8770_1024.targets_bottom
```

Training starts without any warning (at least none that would alert me) but still, the network is not doing anything.

Could somebody guide me through this? @afrozenator maybe? :)

**Comment by stefan-falk** [comment]
## Update

I have now implemented my own version of a multi-problem. I am able to run this but I haven't had time to run enough experiments to see a positive effect yet (if I am at all able to do that).

I simply wasn't able to get the `MultiProblem` to work so I went top-down in a "happy path"-like manner.

### 1. Create `Datasets` and use `sample_from_datasets` for sampling

Essentially what I did was creating a list of `Datasets` for each problem and I use  [`tf.data.experimental.sample_from_datasets`](https://www.tensorflow.org/api_docs/python/tf/data/experimental/sample_from_datasets) in order to randomly sample from datasets according to a constant distribution. 

```python
def dataset(self,
            mode,
            data_dir=None,
            # ..
            ):

    is_training = mode == tf.estimator.ModeKeys.TRAIN

    datasets = list()
    for task in self.tasks:
        task_dataset = task.dataset(
                mode,
                data_dir=data_dir,
                # ..
            )

        task_dataset = task_dataset.map(lambda example: self.add_task_id(task.task_id, example))

        if is_training:
            task_dataset = task_dataset.repeat()

        datasets.append(task_dataset)

    sampled_dataset = tf.data.experimental.sample_from_datasets(
        datasets,
        weights=np.asarray(self.task_weights, dtype=np.float64) / sum(self.task_weights)
    )

    return sampled_dataset
```

### 2. Remove obligatory primary task (language model)

I cannot see a particular reason why a language model is required as primary task. In fact, I think it's more of a burden but please correct me if there is another reason for having this constraint besides the requirement of a shared vocabulary among all tasks. To resolve the shared-vocab issue, I just create all vocabularies of all sub-tasks (sub-problems) and merge them like so:

```python
def get_or_create_vocab(self, data_dir=None, tmp_dir=None, force_get=False) -> SubwordEncoder:
    logger = tf.logging

    vocabs_dir = os.path.join(data_dir, 'shared-vocabs')
    if not tf.gfile.Exists(vocabs_dir):
        tf.gfile.MakeDirs(vocabs_dir)

    vocab_fp = os.path.join(vocabs_dir, self.vocab_filename)

    if tf.gfile.Exists(vocab_fp):
        return SubwordEncoder(vocab_fp)

    # Merge vocabularies

    datasets = self.get_datasets(data_dir)
    reserved_tokens = set()
    subword_tokens = set()
    for dataset in datasets:
        dataset.install()
        encoders = dataset.feature_encoders()
        for (k, encoder) in encoders.items():
            encoder = cast(SubwordEncoder, encoder)
            reserved_tokens.update(encoder.reserved_tokens)
            subword_tokens.update(encoder.subword_tokens)

    final_subwords = RESERVED_TOKENS + sorted(list(subword_tokens.difference(reserved_tokens)))

    logger.info('[%s] - Merged vocabularies; Final vocab size: %s' % (self.name, len(final_subwords)))

    with tf.gfile.Open(vocab_fp, mode='w') as vocab_f:
        for subword in final_subwords:
            vocab_f.write('\'%s\'\n' % subword)

    return SubwordEncoder(vocab_fp)
```

There's probably room for improvement but it works so far.

----

I would like to know whether this looks okay and might have a future. If yes, I might consider creating a pull-request in order to implement a new `MultiProblem` like above. 

Any feedback would be great.

**Comment by stefan-falk** [comment]
I was able to successfully apply my multi-problem implementation to a translation task for German to English (`de2en`) mixing additional tasks `de2es` (to Spanish) and `de2fr` (to French).

![image](https://user-images.githubusercontent.com/43335432/64514699-6da22080-d2eb-11e9-8848-4fa0340eca78.png)

It has to be said that the German `de2en` task had only around 140k sentence pairs. The other datasets where significantly larger. The image shows different outcomes for different dataset weights passed to `tf.data.experimental.sample_from_datasets`.
